src/mlpack/methods/boostmetric/boostmetric_impl.hpp

```diff
@@ -0,0 +1,179 @@
+/**
+ * @file boostmetric_impl.hpp
+ * @author Manish Kumar
+ *
+ * Implementation of BoostMetric class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_BOOSTMETRIC_BOOSTMETRIC_IMPL_HPP
+#define MLPACK_METHODS_BOOSTMETRIC_BOOSTMETRIC_IMPL_HPP
+
+// In case it was not already included.
+#include "boostmetric.hpp"
+
+namespace mlpack {
+namespace boostmetric {
+
+/**
+ * Takes in a reference to the dataset. Copies the data, initializes
+ * all of the member variables and constraint object.
+ */
+template<typename MetricType>
+BOOSTMETRIC<MetricType>::BOOSTMETRIC(const arma::mat& dataset,
+                                     const arma::Row<size_t>& labels,
+                                     const size_t k) :
+    dataset(dataset),
+    labels(labels),
+    k(k),
+    maxIter(500),
+    tolerance(1e-7),
+    wTolerance(1e-5),
+    constraint(dataset, labels, k)
+{ /* nothing to do */ }
+
+// Calculate inner product of two matrices.
+inline double innerProduct(arma::mat& Ar, arma::mat& Z)
```

Comment:
Why not just use `arma::accu(Ar % Z)`?  (Actually `arma::dot()` might do the same thing also, but it's likely you might need a `vectorise()` call there too.)

Reply:
That's much better.